Add scotch

ulmononian opened the pull request:
WW3 requires the `scotch` package for mesh/graph partitioning (see spack-stack https://github.com/NOAA-EMC/spack-stack/issues/465, https://github.com/NOAA-EMC/spack-stack/issues/336).

This simply adds `scotch` to the `common/packages.yaml` and to the ufs-weather-model-static template (though I am not certain this needs to be done here). My understanding is that WW3 needs `scotch`, but I am not sure if this will be added to the UFS common module set (perhaps @JessicaMeixner-NOAA or @MatthewMasarik-NOAA  can comment on that?).

Installs successfully as standalone package and as a part of ufs-weather-model-env (intel & gcc) on Orion.

**For testing purposes, `scotch/7.0.3` has been installed in the Hera spack-stack/1.3.0 Unified Environment here: `/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env`.**

Depends on [NOAA-EMC/spack/260](https://github.com/NOAA-EMC/spack/pull/260).

Would close #465 and #336.

Note: `.gitmodules` pointer and `spack` submodule  will revert back to current spack-stack develop (only modified for testing).

ulmononian committed:
Add ufs-wm-dev template to develop.

ulmononian committed:
Remove ufs-wm-dev template to get fork dev aligned with main dev.

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Update .gitmodules

ulmononian committed:
Update .gitmodules

Revert to EMC spack-stack dev

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

Commit:
Fix .gitmodules branch in develop of fork.

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Merge branch 'NOAA-EMC:develop' into develop

ulmononian committed:
Add scotch to configs/common/packages.yaml.

ulmononian committed:
Add scotch to ufs-wm-static template.

ulmononian committed:
Point to add_scotch branch of NOAA-EMC/spack for testing.

ulmononian committed:
Update gitmodules for testing.

JessicaMeixner-NOAA commented:
@MatthewMasarik-NOAA  and I  actually don't have access to a test to confirm if we'd need this in the ufs-common modules but from earlier comments from @DeniseWorthen I believe this is indeed the case.

JessicaMeixner-NOAA commented:
@MatthewMasarik-NOAA is working to test the new modules with the WW3 standalone regression tests.  @DeniseWorthen can test within the ufs-weather-model coupled framework.

ulmononian commented:
> @MatthewMasarik-NOAA is working to test the new modules with the WW3 standalone regression tests. @DeniseWorthen can test within the ufs-weather-model coupled framework.

thanks for letting me know. it's simple enough for us to add/remove it from the ufs-wm spack/spack-stack configs, so we can play it by ear regarding ufs_common.

climbfuji commented:
Ok, I got the ubuntu CI build going again, and it passed. Will work on the mac CI test tonight and then this is hopefully ready to go.

ulmononian commented:
> Ok, I got the ubuntu CI build going again, and it passed. Will work on the mac CI test tonight and then this is hopefully ready to go.

awesome. thanks @climbfuji. fyi: the PR for the newest scotch version checksum was merged in main spack, so that's done too (https://github.com/spack/spack/pull/37054#event-9062030773).

climbfuji commented:
@ulmononian macOS CI failed building scotch: https://github.com/NOAA-EMC/spack-stack/actions/runs/4755464692/jobs/8449648386?pr=550

climbfuji reviewed (changes requested):
Unfortunately, the macOS build issues need to be sorted out before we can add this to the unified environment.

ulmononian commented:
> @ulmononian macOS CI failed building scotch: https://github.com/NOAA-EMC/spack-stack/actions/runs/4755464692/jobs/8449648386?pr=550

this _looks_ like the same issue with bison version i had on cheyenne. i can't see in the CI log, but if you have access to the full verbose log, it should show which bison version was found. the min. the version needed is 3.4 or newer. i assumed spack would take this into account in the concretize step based on the spec dependencies, but maybe not?

all of that said, can you confirm if the bison version was older than 3.4?

ulmononian commented:
nvm. i see in the concretize log it was bison@2.3. i think this probably caused the failure. can we ensure the bison version is newer than 3.4? i see it is 3.8.2 in common/packages.yaml, but obviously not that version here.

climbfuji commented:
> > @ulmononian macOS CI failed building scotch: https://github.com/NOAA-EMC/spack-stack/actions/runs/4755464692/jobs/8449648386?pr=550
> 
> this _looks_ like the same issue with bison version i had on cheyenne. i can't see in the CI log, but if you have access to the full verbose log, it should show which bison version was found. the min. the version needed is 3.4 or newer. i assumed spack would take this into account in the concretize step based on the spec dependencies, but maybe not?
> 
> all of that said, can you confirm if the bison version was older than 3.4?

Spot on I think. The logs show that an external bison@2.3 was found, and that the concretizer used 2.3 everywhere. If you look at line 57 in https://github.com/ulmononian/NOAA-EMC_spack/compare/4bc1363413325c2ee2f288eecce1134f8c89af58...a088d5d592c05cf83a5033a8eb3bb091dc0e5707#diff-165c5f72581e38ee53bb64e1e9949acf2dea83c7463736c2c48f57b1e9f5cdc4R57 you see that it just wants bison. You'll need to change that line to
```
depends_on("bison@3.4:", type="build")
```
in which case it applies to all scotch versions. Or something like
```
depends_on("bison", when="@:x.y.z", type="build")
depends_on("bison@3.4:", when="@n.m.r:", type="build")
```
if you know what I mean

ulmononian committed:
Update spack hash for bison version upgrade.

ulmononian commented:
@climbfuji thanks for illuminating the path there. i added the `depends_on("bison@3.4:", type="build")` you suggested based on https://gitlab.inria.fr/scotch/scotch/-/issues/21 and updated the spack hash here. hopefully this sorts it.

climbfuji commented:
I had to manually delete stuff from the macos CI buildcache, similar to ubuntu CI, because the PR that updates spack from the authoritative repo adds packages (as dependencies) that do not exist in the spack base for your PR. CI is running now:

https://github.com/NOAA-EMC/spack-stack/actions/runs/4761065096/jobs/8470656997

climbfuji commented:
Yay! It worked :-)

climbfuji reviewed (changes requested):
See https://github.com/NOAA-EMC/spack/pull/260#pullrequestreview-1395910610

MatthewMasarik-NOAA commented:
I've been working on testing the `hera` installation.  This is the path I'm pointing to there:
```
export SCOTCH_PATH=/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/intel/2021.5.0/scotch-7.0.3-a5gaddh
```
An initial find is that the spack build produces `*.so` files, whereas the hpc-stack (and our wave team builds) produce `*.a` files.  I'm investigating this more to see how the WW3 build behaves.

ulmononian commented:
> I've been working on testing the `hera` installation. This is the path I'm pointing to there:
> 
> ```
> export SCOTCH_PATH=/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/intel/2021.5.0/scotch-7.0.3-a5gaddh
> ```
> 
> An initial find is that the spack build produces `*.so` files, whereas the hpc-stack (and our wave team builds) produce `*.a` files. I'm investigating this more to see how the WW3 build behaves.

thanks so much for testing this out! two things to note: spack-stack uses shared parallelio and static esmf libraries, but as i am not sure how ww3 is built, i am not sure if this is of any consequence to you seeing .so files built rather than .a files.

MatthewMasarik-NOAA commented:
> thanks so much for testing this out! two things to note: spack-stack uses shared parallelio and static esmf libraries, but as i am not sure how ww3 is built, i am not sure if this is of any consequence to you seeing .so files built rather than .a files.

@ulmononian, sure thing!  I'm not sure of how ww3 will react either, I'll keep you posted when I know.

MatthewMasarik-NOAA commented:
@ulmononian, I altered the WW3 cmake build to find the `*.so` files that are present, but found that the spack build does not produce the same (read analogous *.so) files found in our Wave group build, or the hpc-stack build.  So, the build is crashing now because it can't find those files.

#### Wave group build
```bash
[Matthew.Masarik@hfe09 scotch_testing]$ ls -1 /scratch1/NCEPDEV/climate/Matthew.Masarik/waves/opt/hpc-stack/scotch-v7.0.3/install/lib
cmake
libesmumps.a
libptesmumps.a
libptscotch.a
libptscotcherr.a
libptscotcherrexit.a
libptscotchparmetisv3.a
libscotch.a
libscotcherr.a
libscotcherrexit.a
libscotchmetisv3.a
libscotchmetisv5.a
```

#### Hpc-stack build
```bash
[Matthew.Masarik@hfe09 scotch_testing]$ ls -1 /scratch1/NCEPDEV/nems/role.epic/hpc-stack/libs/intel-2022.1.2/intel-2022.1.2/impi-2022.1.2/scotch/7.0.3/lib
cmake
libesmumps.a
libptesmumps.a
libptscotch.a
libptscotcherr.a
libptscotcherrexit.a
libptscotchparmetisv3.a
libscotch.a
libscotcherr.a
libscotcherrexit.a
libscotchmetisv3.a
libscotchmetisv5.a
```

#### Spack-stack build
```
[Matthew.Masarik@hfe09 scotch_testing]$ ls -1 /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/intel/2021.5.0/scotch-7.0.3-a5gaddh/lib
cmake
libptscotcherrexit.so
libptscotcherr.so
libptscotch.so
libptscotch.so.7.0.3
libscotcherrexit.so
libscotcherr.so
libscotch.so
libscotch.so.7.0.3
```

ulmononian commented:
> @ulmononian, I altered the WW3 cmake build to find the `*.so` files that are present, but found that the spack build does not produce the same (read analogous *.so) files found in our Wave group build, or the hpc-stack build. So, the build is crashing now because it can't find those files.
> #### Wave group build
> 
> ```shell
> [Matthew.Masarik@hfe09 scotch_testing]$ ls -1 /scratch1/NCEPDEV/climate/Matthew.Masarik/waves/opt/hpc-stack/scotch-v7.0.3/install/lib
> cmake
> libesmumps.a
> libptesmumps.a
> libptscotch.a
> libptscotcherr.a
> libptscotcherrexit.a
> libptscotchparmetisv3.a
> libscotch.a
> libscotcherr.a
> libscotcherrexit.a
> libscotchmetisv3.a
> libscotchmetisv5.a
> ```
> 
> #### Hpc-stack build
> 
> ```shell
> [Matthew.Masarik@hfe09 scotch_testing]$ ls -1 /scratch1/NCEPDEV/nems/role.epic/hpc-stack/libs/intel-2022.1.2/intel-2022.1.2/impi-2022.1.2/scotch/7.0.3/lib
> cmake
> libesmumps.a
> libptesmumps.a
> libptscotch.a
> libptscotcherr.a
> libptscotcherrexit.a
> libptscotchparmetisv3.a
> libscotch.a
> libscotcherr.a
> libscotcherrexit.a
> libscotchmetisv3.a
> libscotchmetisv5.a
> ```
> 
> #### Spack-stack build
> 
> ```
> [Matthew.Masarik@hfe09 scotch_testing]$ ls -1 /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/intel/2021.5.0/scotch-7.0.3-a5gaddh/lib
> cmake
> libptscotcherrexit.so
> libptscotcherr.so
> libptscotch.so
> libptscotch.so.7.0.3
> libscotcherrexit.so
> libscotcherr.so
> libscotch.so
> libscotch.so.7.0.3
> ```

hi @MatthewMasarik-NOAA. i am doing another install of `scotch@7.0.3` with these variants `+mpi+metis~shared` (mentioned by @JessicaMeixner-NOAA [here](https://github.com/NOAA-EMC/spack-stack/issues/465#issuecomment-1412704997)). would you be willing to test this version when its ready?

MatthewMasarik-NOAA commented:
> hi @MatthewMasarik-NOAA. i am doing another install of `scotch@7.0.3` with these variants `+mpi+metis~shared` (mentioned by @JessicaMeixner-NOAA [here](https://github.com/NOAA-EMC/spack-stack/issues/465#issuecomment-1412704997)). would you be willing to test this version when its ready?

@ulmononian  Yes!  Absolutely.  When it's ready, let me know and I'll test it.

ulmononian commented:
> > hi @MatthewMasarik-NOAA. i am doing another install of `scotch@7.0.3` with these variants `+mpi+metis~shared` (mentioned by @JessicaMeixner-NOAA [here](https://github.com/NOAA-EMC/spack-stack/issues/465#issuecomment-1412704997)). would you be willing to test this version when its ready?
> 
> @ulmononian Yes! Absolutely. When it's ready, let me know and I'll test it.

thank you!! it is installed now. you can load it just like you loaded before (the old installation was replaced with this variant).

ulmononian commented:
@MatthewMasarik-NOAA just to note: the installation `lib` contents for the `+mpi+metis~shared` variant are:

```
cmake libptscotcherr.a  libptscotchparmetisv3.a  libscotcherr.a libscotchmetisv3.a
libptscotch.a  libptscotcherrexit.a  libscotch.a libscotcherrexit.a  libscotchmetisv5.a
```

whereas the initial installation `lib` directory contains:

```
cmake libptscotcherr.so  libptscotch.so.7.0.3 libscotcherr.so  libscotch.so.7.0.3
libptscotcherrexit.so  libptscotch.so libscotcherrexit.so libscotch.so
```

MatthewMasarik-NOAA commented:
@ulmononian that's perfect.  Thanks for the update.  I'll let you know how testing goes

MatthewMasarik-NOAA commented:
@ulmononian  with this install I was able to build WW3 OK.  Though when I tried a test run with it, it crashed, similar to how the hpc-stack scotch crashes.  

I tested it with the most recent version of WW3 develop, which now has a fix from @DeniseWorthen (NOAA-EMC/WW3/pull/1001), she found working with SCOTCH compiled with GNU. 

I am able to run the same tests by using a SCOTCH build I had compiled.  For this, the job card script is the same, with only the modules and SCOTCH_PATH environment variable being different.  I've shown below for each, the job card setup and the corresponding `module list` from the job output log file.

The crash is in a spot we've seen it before with SCOTCH -- during the WW3 model initialization.  This could be related to the issue of SCOTCH needing updated gnu headers for compilation.  In the build instructions for `intel` we module load `gnu/9.2.0` at the end as a fix for this.  I'm not sure how it would be accounted for with spack-stack.

## I. spack-stack SCOTCH install
**job card environment**
```bash
  module purge
  module use /scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles
  module load miniconda/3.9.12
  module load ecflow/5.5.3
  module load mysql/8.0.31
  module use /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/modulefiles/Core
  module load stack-intel/2021.5.0
  module load stack-intel-oneapi-mpi/2021.5.1
  module load stack-python/3.9.12

  module load cmake/3.20.1
  module load hpc/1.2.0
  module load hpc-intel/2022.1.2
  module load hpc-impi/2022.1.2
  module load libpng/1.6.37
  module load zlib/1.2.11
  module load jasper/2.0.25
  module load hdf5/1.10.6
  module load netcdf/4.7.4
  module load bacio/2.4.1
  module load g2/3.4.5
  module load w3emc/2.9.2
  module load esmf/8.3.0b09
  module load scotch/7.0.3
  export SCOTCH_PATH=/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/intel/2021.5.0/scotch-7.0.3-r2wrzvc
```
**job output - module list**
```
Currently Loaded Modules:
  1) ecflow/5.5.3                     13) libjpeg/2.1.0
  2) mysql/8.0.31                     14) jasper/2.0.32
  3) intel/2022.1.2                   15) g2/3.4.5
  4) stack-intel/2021.5.0             16) w3emc/2.9.2
  5) impi/2022.1.2                    17) curl/7.29.0
  6) stack-intel-oneapi-mpi/2021.5.1  18) pkg-config/0.27.1
  7) miniconda/3.9.12                 19) hdf5/1.14.0
  8) stack-python/3.9.12              20) zstd/1.5.2
  9) cmake/3.20.1                     21) netcdf-c/4.9.2
 10) zlib/1.2.13                      22) netcdf-fortran/4.6.0
 11) libpng/1.6.37                    23) esmf/8.3.0b09
 12) bacio/2.4.1                      24) scotch/7.0.3
```
<br>

## II. Code mgmt SCOTCH install
**job card environment**
```bash
module purge
  module load cmake/3.20.1
  module load intel/2022.1.2
  module load impi/2022.1.2
  module use  /scratch1/NCEPDEV/nems/role.epic/hpc-stack/libs/intel-2022.1.2/modulefiles/stack
  module load hpc/1.2.0
  module load hpc-intel/2022.1.2
  module load hpc-impi/2022.1.2
  module load libpng/1.6.37
  module load zlib/1.2.11
  module load jasper/2.0.25
  module load hdf5/1.10.6
  module load netcdf/4.7.4
  module load bacio/2.4.1
  module load g2/3.4.5
  module load w3emc/2.9.2
  module load esmf/8.3.0b09
  export SCOTCH_PATH=/scratch1/NCEPDEV/climate/Matthew.Masarik/waves/opt/hpc-stack/scotch-v7.0.3/install
```

**job output - module list**
```
Currently Loaded Modules:
  1) cmake/3.20.1         6) hpc-impi/2022.1.2  11) netcdf/4.7.4
  2) hpc/1.2.0            7) libpng/1.6.37      12) bacio/2.4.1
  3) intel/2022.1.2       8) zlib/1.2.11        13) g2/3.4.5
  4) hpc-intel/2022.1.2   9) jasper/2.0.25      14) w3emc/2.9.2
  5) impi/2022.1.2       10) hdf5/1.10.6        15) esmf/8.3.0b09
```

ulmononian commented:
@MatthewMasarik-NOAA thanks for the continued testing and this info. ultimately, if these gnu headers are needed for compilation, it will need to be handled in https://github.com/ulmononian/NOAA-EMC_spack/blob/add_scotch/var/spack/repos/builtin/packages/scotch/package.py. since loading gnu often results in conflicts when using intel compilers, it seems like a special, non-conflicting gnu module would need to be added to all machines; from my understanding this has only be done officially on orion (i created one for testing on cheyenne). the spack build would then need to be able to load these special gnu modules.

has there been any success running ww3 tests with a spack-built scotch in the past by the ww3 team? i am only aware of it being _built_ successfully as mentioned [here ](https://github.com/NOAA-EMC/spack-stack/issues/465#issuecomment-1412704997)by @JessicaMeixner-NOAA.

if your team wants to look at this build recipe  https://github.com/ulmononian/NOAA-EMC_spack/blob/add_scotch/var/spack/repos/builtin/packages/scotch/package.py and perhaps identify a way to ensure the proper gnu headers are loaded, we can add this to the authoritative repo and then re-rerun the installs. gnu can be added as a dependency to this py script, but we would have to ensure it loads the gnu that is meant to not conflict with intel (when intel is the compiler).

AlexanderRichert-NOAA commented:
@MatthewMasarik-NOAA do you know what specific setting(s) scotch needs from the gcc modules? For some systems, as part of our spack-stack configuration we already update $PATH, $CPATH, and $LD_LIBRARY_PATH to include a GNU installation for the Intel compiler, so I wonder if we can use that to our advantage here (and basically populate those in the `stack-intel` module to avoid the issue of conflicting compiler modules).

MatthewMasarik-NOAA commented:
> has there been any success running ww3 tests with a spack-built scotch in the past by the ww3 team? i am only aware of it being _built_ successfully as mentioned [here ](https://github.com/NOAA-EMC/spack-stack/issues/465#issuecomment-1412704997)by @JessicaMeixner-NOAA.

@ulmononian, to my knowledge there has not been a successful run by the ww3 team with a spack-built scotch.  The build @JessicaMeixner-NOAA referred to was on github CI, so not in hpc.

MatthewMasarik-NOAA commented:
> @MatthewMasarik-NOAA do you know what specific setting(s) scotch needs from the gcc modules? For some systems, as part of our spack-stack configuration we already update $PATH, $CPATH, and $LD_LIBRARY_PATH to include a GNU installation for the Intel compiler, so I wonder if we can use that to our advantage here (and basically populate those in the `stack-intel` module to avoid the issue of conflicting compiler modules).

@AlexanderRichert-NOAA, I believe scotch just needs newer header files than exist with the system gcc.  Possibly updating the paths as you described would achieve that?  @aerorahul had looked at this with @JessicaMeixner-NOAA and I a little bit ago.  He may be able to confirm or correct the record here.

JessicaMeixner-NOAA commented:
So I had to refresh my memory as to what the requirements for GNU were, so I went back to this: 

https://gitlab.inria.fr/scotch/scotch/-/issues/21 which mostly points to the Bison version. 

The GNU is spoken about in this issue: https://gitlab.inria.fr/scotch/scotch/-/issues/19 where it's about which "C standard version they are compiling with".     

I do believe we build and ran with spack on orion at one point, but we were testing so many different things and were having other issues that I don't remember what succeeded and what did not.

climbfuji commented:
I didn't follow the conversation but saw a mentioning of C++ standard. Right now we are using C++-14 in places like boost, but we are hoping to update to C++-17 in the near future.

AlexanderRichert-NOAA commented:
Hmm... It looks like in `/work/noaa/epic-ps/role-epic-ps/spack-stack/spack-stack-1.3.0/envs/unified-env/install/modulefiles/Core/stack-intel/2022.0.2.lua`, it's already updating PATH, LD_LIBRARY_PATH, and CPATH. In the gcc modules, the only relevant variables I see getting modified are PATH and LD_LIBRARY_PATH, so I'm surprised that doesn't take care of it... @MatthewMasarik-NOAA have you tried with gcc version more recent than 8.3.0? Mostly I ask because there are some subtle differences in the LD_LIBRARY_PATH settings between the gcc modules, so if it _didn't_ work with, say, gcc 10 or 11, that could help narrow things down.

MatthewMasarik-NOAA commented:
@AlexanderRichert-NOAA, for non-spack installs we have used gcc/9.2.0 on hera and gcc/10.2.0 on orion and acorn.  Is that what you're interested in?   I don't know what has been tried within spack builds.

AlexanderRichert-NOAA commented:
Okay yeah that answers my question (which was about Orion except I forgot to mention it). In that case I'm surprised it's not already working...

MatthewMasarik-NOAA commented:
@AlexanderRichert-NOAA okay, hmm.  It's hard to know at this point if the issue is related to gcc version.

On another note, looking at the build recipe @ulmononian posted, there is one question I had, and a comment.

In the lines, https://github.com/ulmononian/NOAA-EMC_spack/blob/ade0b41863d38739f8b94354e720b927b4562cd5/var/spack/repos/builtin/packages/scotch/package.py#L122-L123, does the `-DINTSIZE=64` option end up getting used for the compiles?  I saw `int64` which is used in the `if` statement has a default set to False in the `variant( )` call above, so it looks like 'no'?  If that is the case, good.  We do not want to build with `-DINTSIZE=64`, currently.

Then the comment I had related to a TODO in a line right above the other two, https://github.com/ulmononian/NOAA-EMC_spack/blob/ade0b41863d38739f8b94354e720b927b4562cd5/var/spack/repos/builtin/packages/scotch/package.py#L120.  We should disable THREADS, as well as MPI_THREAD_MULTIPLE.  This shouldn't be necessary on hera (where I've tested so far), though on orion it was necessary to disable `THREADS`.

MatthewMasarik-NOAA commented:
Hi @ulmononian @AlexanderRichert-NOAA, I wanted to check in on this to see if there is an updated install I should test?

ulmononian commented:
> Hi @ulmononian @AlexanderRichert-NOAA, I wanted to check in on this to see if there is an updated install I should test?

Hi @MatthewMasarik-NOAA. Apologies, but I have not had time to work on this recently. I will try to get back to it soon.

ulmononian commented:
> Hi @ulmononian @AlexanderRichert-NOAA, I wanted to check in on this to see if there is an updated install I should test?

@MatthewMasarik-NOAA  i've added two variants to the build recipe (to build w/  `THREADS` and `MPI_THREAD_MULTIPLE` set to `OFF`). i know you mentioned this was probably not required (so probably not the issue) on hera, but maybe we will get lucky. do you mind testing with this again (scotch has been updated but the module is just `scotch/7.0.3`): `/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env`?

MatthewMasarik-NOAA commented:
Hi @ulmononian, thank you for providing this new scotch build.  Yes, I'll test it and let you know the outcome.

ulmononian commented:
@AlexanderRichert-NOAA perhaps you can weigh in on this comment from a scotch developer regarding the gnu issue:

"To my understanding, this is not an issue of gnu or intel compiler versions per se, but of which C standard version they are referring to when processing the source code files. Indeed, PRIu64 and its likes were inserted in C99 : https://en.cppreference.com/w/c/types/integer. Hence, in order to compile Scotch without errors, one has to make sure the compiler accepts the C99 standard,e.g., by using the "`-std=c99`" flag in gcc. Include files should already have the proper information, but what matters is to provide the adequate #define's to make this information accessible to the compilers."

i loaded up the stack-intel on hera and did ` gcc -std=c99 -dM -E - < /dev/null | grep __STDC_VERSION__` which shows:
`#define __STDC_VERSION__ 199901L`, so it seems like the C99 is accepted by the `gnu/9.2` that is included in the stack-intel compiler modulefile (`/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/modulefiles/Core/stack-intel/2021.5.0.lua`).

based on this, i'm not sure if this it is indeed a gnu issue causing the ww3 init failures on hera right now (cc @MatthewMasarik-NOAA )

AlexanderRichert-NOAA commented:
My understanding is that the WW3 issues have to do with runtime issue related to the graph infrastructure, as opposed to SCOTCH buildtime issues. @MatthewMasarik-NOAA does that sound right. @ulmononian have you tested building SCOTCH with Intel? As far as I know all the intel versions we care about have c99 support so I don't think that should be an issue.

MatthewMasarik-NOAA commented:
@ulmononian @AlexanderRichert-NOAA, Alex is correct that the issues are seen during runtime, specifically during the model initialization where SCOTCH is called and the graph partitioning occurs.  Since we had a work-around that did the trick (loading `gnu` at the end of the `intel` loads), compiling the C portions with `-std=99` has not been investigated.

ulmononian commented:
> My understanding is that the WW3 issues have to do with runtime issue related to the graph infrastructure, as opposed to SCOTCH buildtime issues. @MatthewMasarik-NOAA does that sound right. @ulmononian have you tested building SCOTCH with Intel? As far as I know all the intel versions we care about have c99 support so I don't think that should be an issue.

so the version i am building for @MatthewMasarik-NOAA to test is w/ intel. i was asking about the gnu c99 comment only in regard to the gnu 9.2 that is added to the intel modulefile paths. i wanted to make sure it had the appropriate settings. everything compiles "successfully", so that is not an issue. but the runtime ww3 init failure suggest to me something is poorly configured for the compilation. or the mpi init step in the model needs to be changed perhaps.

MatthewMasarik-NOAA commented:
> @MatthewMasarik-NOAA i've added two variants to the build recipe (to build w/ `THREADS` and `MPI_THREAD_MULTIPLE` set to `OFF`). i know you mentioned this was probably not required (so probably not the issue) on hera, but maybe we will get lucky. do you mind testing with this again (scotch has been updated but the module is just `scotch/7.0.3`): `/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env`?
<br>

@ulmononian, I haven't been successful getting all the modules loaded when I'm using this latest path.

Here's what I'm doing on hera:
```
module purge
module load cmake/3.20.1
module load intel/2022.1.2
module load impi/2022.1.2
```
then
```
module use /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env
```
at this point when I do a `module avail` I get a pretty long list, with some truncated output shown here
```
$ module avail
   .
   .
   intel/2021.5.0/json/3.10.5
   intel/2021.5.0/krb5/1.15.1
   intel/2021.5.0/landsfcutil/2.4.1
   intel/2021.5.0/libbsd/0.11.5
   intel/2021.5.0/libidn2/2.3.0
   intel/2021.5.0/libjpeg/2.1.0
   intel/2021.5.0/libmd/1.0.4
   intel/2021.5.0/libpng/1.6.37
   .
   .
```
The next package we want to load for WW3 is `libpng/1.6.37`, the last line in the code snippet above shows this.  I tried loading it the two sensible ways, shown here
```
[Matthew.Masarik@hfe04 regtests(develop)]$ module load intel/2021.5.0/libpng/1.6.37
Lmod has detected the following error:  These module(s) or extension(s) exist but cannot be loaded as requested: "zlib/1.2.13"
   Try: "module spider zlib/1.2.13" to see how to load the module(s).

```
and
```
[Matthew.Masarik@hfe04 regtests(develop)]$ module load libpng/1.6.37
Lmod has detected the following error:  These module(s) or extension(s) exist but cannot be loaded as requested: "libpng/1.6.37"
   Try: "module spider libpng/1.6.37" to see how to load the module(s).
```

I also tried being more specific with the path specified in the `module use` statement.  Including
```
module use /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/modulefiles/Core/stack-intel
```
though for any of these attempts I also go the same output when i ran a 'module avail'.  

Can you see what might be going wrong here?

climbfuji commented:
Please follow the instructions in https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html#noaa-rdhpcs-hera for loading the modules - let me know if this works.

MatthewMasarik-NOAA commented:
> Please follow the instructions in https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html#noaa-rdhpcs-hera for loading the modules - let me know if this works.

ah! right.  Thank you @climbfuji

ulmononian commented:
@MatthewMasarik-NOAA 

please do this first (for intel):

```
module purge
module use /scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles
module load miniconda/3.9.12
module load ecflow/5.5.3
module load mysql/8.0.31

module use /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/modulefiles/Core
module load stack-intel/2021.5.0
module load stack-intel-oneapi-mpi/2021.5.1
module load stack-python/3.9.12
```

then, based on the modules you were loading in your comment [here](https://github.com/JCSDA/spack-stack/pull/550#issuecomment-1523872426), please do:

```
  module load libpng/1.6.37
  module load zlib/1.2.13
  module load jasper/2.0.32
  module load hdf5/1.14.0
  module load netcdf-c/4.9.2
  module load netcdf-fortran/4.6.0
  module load bacio/2.4.1
  module load g2/3.4.5
  module load w3emc/2.9.2
  module load esmf/8.3.0b09
  module load scotch/7.0.3
```

some of these versions differ from what you were loading, but they are the versions provided in spack-stack/1.3.0. obviously you do not need to load all of these if your module env has changed. please let me know if this works for you to at least get to the point where you can compile/start to run ww3.

MatthewMasarik-NOAA commented:
> @MatthewMasarik-NOAA
> 
> please do this first (for intel):
> 
> ```
> module purge
> module use /scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles
> module load miniconda/3.9.12
> module load ecflow/5.5.3
> module load mysql/8.0.31
> 
> module use /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.3.0/envs/unified-env/install/modulefiles/Core
> module load stack-intel/2021.5.0
> module load stack-intel-oneapi-mpi/2021.5.1
> module load stack-python/3.9.12
> ```
> 
> then, based on the modules you were loading in your comment [here](https://github.com/JCSDA/spack-stack/pull/550#issuecomment-1523872426), please do:
> 
> ```
>   module load libpng/1.6.37
>   module load zlib/1.2.13
>   module load jasper/2.0.32
>   module load hdf5/1.14.0
>   module load netcdf-c/4.9.2
>   module load netcdf-fortran/4.6.0
>   module load bacio/2.4.1
>   module load g2/3.4.5
>   module load w3emc/2.9.2
>   module load esmf/8.3.0b09
>   module load scotch/7.0.3
> ```
> 
> some of these versions differ from what you were loading, but they are the versions provided in spack-stack/1.3.0. obviously you do not need to load all of these if your module env has changed. please let me know if this works for you to at least get to the point where you can compile/start to run ww3.

Perfect.  Thank you, @ulmononian!!

ulmononian commented:
@MatthewMasarik-NOAA any luck with this new installation by chance?

MatthewMasarik-NOAA commented:
> @MatthewMasarik-NOAA any luck with this new installation by chance?

@ulmononian, I got pulled away from my attempts to test this yesterday. I just got it set up, and the tests have been going for 20mins now.  I'll know the outcome ~2hrs.  I'll report back shortly after

MatthewMasarik-NOAA commented:
@ulmononian, good news, all the tests went great!  I loaded the modules you listed and WW3 built and ran fine (without even needing to set the SCOTCH_PATH env parameter).  Thank you for re-doing the build!

climbfuji commented:
Great! Thanks so much for this update. I'll work with @ulmononian to get this and other PRs merged this afternoon

MatthewMasarik-NOAA commented:
> Great! Thanks so much for this update. I'll work with @ulmononian to get this and other PRs merged this afternoon

Sure thing.  Thank you!

ulmononian commented:
> @ulmononian, good news, all the tests went great! I loaded the modules you listed and WW3 built and ran fine (without even needing to set the SCOTCH_PATH env parameter). Thank you for re-doing the build!

awesome! very glad this finally worked. i assume we should just go with the threading variants i added, so i will update that over at https://github.com/JCSDA/spack/pull/260. thanks for testing @MatthewMasarik-NOAA.

MatthewMasarik-NOAA commented:
> awesome! very glad this finally worked. i assume we should just go with the threading variants i added, so i will update that over at [JCSDA/spack#260](https://github.com/JCSDA/spack/pull/260). thanks for testing @MatthewMasarik-NOAA.

Yes, I agree we should go with the threading variants you added.  thanks @ulmononian

ulmononian committed:
Merge branch 'develop' of https://github.com/JCSDA/spack-stack into feature/add_scotch

ulmononian committed:
Update spack submodule.

ulmononian committed:
Merge branch 'develop' of https://github.com/JCSDA/spack-stack into feature/add_scotch

ulmononian committed:
Point to scotch spack branch and submodule.

ulmononian committed:
Update spack submodule and .gitmodules.

ulmononian committed:
Update spack.

climbfuji reviewed (approved):
scotch was tested successfully on Hera
